### ISSUE TITLE
/transaction/announce POST Request for v0.6.98 nodes fail with CORS error, but for v0.6.97 nodes succeed

### ISSUE THREAD

**Issue by YasunoriMATSUOKA** [NemProject/nis#12]
## What happend?

On https://nemnodes.org/nodes/, in NIS1 v0.6.98 nodes, I tried to find the node with the latest block height, set the domain or IP address of that node to Wallet, and send a transaction to send XEM without a message, with the following result Failed to send the transaction.

However, when I set the IP address of NIS1 v0.6.97 node to Wallet and sent the same transaction, it succeeded.

![image](https://user-images.githubusercontent.com/55148923/140649901-a3bd7a37-dfcb-4b9d-9b3d-7d72c4895787.png)

![image](https://user-images.githubusercontent.com/55148923/140666039-2606b654-8dd1-4b1c-99b8-3f036a2345b6.png)

Error message

```
Access to fetch at 'http://xem22.allnodes.me:7890/transaction/announce' from origin 'null' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.
request.js:150 POST http://xem22.allnodes.me:7890/transaction/announce net::ERR_FAILED
ClientRequest._onFinish @ request.js:150
(anonymous) @ request.js:62
EventEmitter.emit @ events.js:78
finishMaybe @ _stream_writable.js:630
afterWrite @ _stream_writable.js:492
afterTick @ index.js:41
Item.run @ browser.js:153
drainQueue @ browser.js:123
setTimeout (async)
runTimeout @ browser.js:41
process.nextTick @ browser.js:143
exports.setImmediate @ main.js:57
Request.init @ request.js:525
RP$initInterceptor @ request2.js:45
Request @ request.js:130
request @ index.js:54
(anonymous) @ send.js:26
send @ send.js:25
announce @ transaction.js:33
send @ send.js:49
_transact @ wallet.service.js:279
transact @ wallet.service.js:294
send @ transferTransaction.controller.js:182
fn @ VM353:4
callback @ angular.js:26808
$eval @ angular.js:18017
$apply @ angular.js:18117
(anonymous) @ angular.js:26813
defaultHandlerWrapper @ angular.js:3617
eventHandler @ angular.js:3605
plumbing.js:87 Uncaught (in promise) RequestError {name: 'RequestError', message: 'TypeError: Failed to fetch', cause: TypeError: Failed to fetch
    at module.exports.ClientRequest._onFinish (file:///C:/Users/yasun/NE…, error: TypeError: Failed to fetch
    at module.exports.ClientRequest._onFinish (file:///C:/Users/yasun/NE…, options: {…}, …}
```

## Environment

- OS: Windows 10
- Browser: Chrome: 94.0.4606.81
- Wallet: NEM Wallet Universal v2.6.1

## Opinion

In NIS1 v0.6.98, it seemed that the preflight OPTIONS Request generated at /transaction/announce POST Request was getting a 500 Server Error.
Isn't there some problem with the OPTIONS Method processing in NIS1 v0.6.98?

(I have not tested this, but) I am concerned that this issue may be causing all transactions that set up delegated harvesting for nodes that have been upgraded to NIS1 v0.6.98 to fail.
I think this will have a big negative impact on the ongoing decision to the community on whether or not to do a Harlock hard fork.
Please confirm. Thank you!

**Comment by cryptoBeliever** [comment]
Example OPTION request with 500 response code sent by browser:
```
curl -v 'http://xem22.allnodes.me:7890/transaction/announce' \
  -X 'OPTIONS' \
  -H 'Connection: keep-alive' \
  -H 'Accept: */*' \
  -H 'Access-Control-Request-Method: POST' \
  -H 'Access-Control-Request-Headers: content-type' \
  -H 'Origin: null' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Accept-Language: en-GB,en;q=0.9,pl-PL;q=0.8,pl;q=0.7,en-US;q=0.6' \
  --compressed \
  --insecure
```
Exception on NIS1 side:

```
org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.ClassCastException: org.springframework.web.servlet.handler.AbstractHandlerMapping$PreFlightHandler cannot be cast to org.springframework.web.method.HandlerMethod
    at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:982)
    at org.springframework.web.servlet.FrameworkServlet.doOptions(FrameworkServlet.java:908)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:716)
    at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
    at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1459)
    at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
    at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1626)
    at org.nem.deploy.server.AbstractNemServletContextListener$1.doFilter(AbstractNemServletContextListener.java:123)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
    at org.eclipse.jetty.servlets.GzipFilter.doFilter(GzipFilter.java:50)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
    ....
    at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
    at org.eclipse.jetty.server.Server.handle(Server.java:512)
    at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
    at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
    at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:137)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
    at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: org.springframework.web.servlet.handler.AbstractHandlerMapping$PreFlightHandler cannot be cast to org.springframework.web.method.HandlerMethod
    at org.nem.nis.controller.interceptors.LocalHostInterceptor.preHandle(LocalHostInterceptor.java:30)
    at org.springframework.web.servlet.HandlerExecutionChain.applyPreHandle(HandlerExecutionChain.java:139)
    at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:962)
    at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
    at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
    ... 43 more

```

Most probably caused by changes in Spring 4.2 regarding CORS support (https://docs.spring.io/spring-framework/docs/4.2.x/spring-framework-reference/html/cors.html). In version 0.6.98 we using spring 4.3.30, in previous version 0.6.97 it was used spring 4.1.6. On spring 4.1.6 interceptors like `LocalHostInterceptor` was not called for OPTIONS requests.

================

Potential solution: Treat handlers other that HandlerMathod as public:

```
@Override
	public boolean preHandle(final HttpServletRequest request, final HttpServletResponse response, final Object handler) throws Exception {

		if (!isTrustedApi(handler) || this.localHostDetector.isLocal(request)) {
			return true;
		}

		final String message = String.format("remote %s attempted to call local %s", request.getRemoteAddr(), request.getRequestURI());
		LOGGER.warning(message);
		throw new UnauthorizedAccessException(message);
	}

	private boolean isTrustedApi(final Object handler) {
		if (handler instanceof HandlerMethod) {
			final HandlerMethod handlerMethod = (HandlerMethod) handler;
			final Method method = handlerMethod.getMethod();
			return method.isAnnotationPresent(TrustedApi.class);
		} else {
			return false;
		}
	}

```

**Comment by YasunoriMATSUOKA** [comment]
You may already be aware of this, but here is a summary of the phenomenon that I have been able to confirm here.

- From NEM Wallet Universal to NIS1 v0.6.98 -> Failure
- From NEM Wallet Universal to NIS1 v0.6.97 -> Success
- From NEM Wallet Windows to NIS1 v0.6.98 -> Success
- From NEM Wallet Windows to NIS1 v0.6.97 -> Success

The above phenomenon was reproduced in the following environment. (But, I have not verified all the patterns. This is because I thought it was too much and that the problem itself was sorted out.)

- NIS1 v0.6.97 ... BinanceStaking001 - BinanceStaking030
- NIS1 v0.6.98 ... nem-sakura-16.next-web-technology.com (My NIS1 node)

If it is difficult to solve the short-term problem with NEM Wallet Universal and NIS1, I think it would be better to make an announcement to the community to use other types of desktop wallets instead of Universal Wallet. How do you think about it?

Thank you for your cooperation.

**Comment by gimre-xymcity** [comment]
@YasunoriMATSUOKA can you confirm it's fixed in v0.6.99?

**Comment by YasunoriMATSUOKA** [comment]
@gimre-xymcity 
Yes, I can successfully announce the following transaction from NEM Wallet Universal to sn1.tamami-foundation.org in NIS1 v0.6.99.

https://explorer.nemtool.com/#/s_tx?hash=b5a6fdb098623c603ce79da6655ac1e4ab400eada2c6340e2bb2753c42960095

Thank you for fixing the issue.
